File: examples/checks/complex_matrix_stability.py
# Independent Python checks for the complex_matrix_stability example.
from __future__ import annotations

import math
import re

from .common import run_checks


def modulus(z: dict) -> float:
    return math.hypot(float(z["re"]), float(z["im"]))


def classify(radius: float) -> str:
    if radius < 1.0:
        return "damped"
    if abs(radius - 1.0) <= 1e-12:
        return "marginally stable"
    return "unstable"


def parse_rows(answer: str) -> dict[str, tuple[float, str]]:
    rows = {}
    for name, radius, label in re.findall(r"^([A-Za-z0-9_]+) : spectral radius ([0-9.]+) -> ([^\n]+)$", answer, flags=re.MULTILINE):
        rows[name] = (float(radius), label.strip())
    return rows


def run(ctx):
    data = ctx.load_input()
    rows = parse_rows(ctx.answer)
    radii = {
        matrix["name"]: max(modulus(z) for z in matrix["diagonal"])
        for matrix in data["matrices"]
    }
    z = complex(float(data["sampleProduct"]["z"]["re"]), float(data["sampleProduct"]["z"]["im"]))
    w = complex(float(data["sampleProduct"]["w"]["re"]), float(data["sampleProduct"]["w"]["im"]))
    scale = float(data["scale"])

    checks = [
        ("diagonal entries are used as the eigenvalues", set(radii) == {m["name"] for m in data["matrices"]}),
        ("A_unstable has independently recomputed spectral radius 2", abs(radii["A_unstable"] - 2.0) <= 1e-12 and classify(radii["A_unstable"]) == "unstable"),
        ("A_stable has spectral radius exactly 1 and is marginal", abs(radii["A_stable"] - 1.0) <= 1e-12 and classify(radii["A_stable"]) == "marginally stable"),
        ("A_damped has spectral radius 0 and is damped", radii["A_damped"] == 0.0 and classify(radii["A_damped"]) == "damped"),
        ("reported matrix classes and radii match recomputation", set(rows) == set(radii) and all(abs(rows[name][0] - radii[name]) <= 1e-12 and rows[name][1] == classify(radii[name]) for name in rows)),
        ("squared modulus of z*w equals product of squared moduli", abs(abs(z * w) ** 2 - (abs(z) ** 2 * abs(w) ** 2)) <= 1e-12),
        ("scaling a matrix by 2 multiplies spectral-radius-squared by 4", abs((scale * radii["A_unstable"]) ** 2 - (scale ** 2) * radii["A_unstable"] ** 2) <= 1e-12),
    ]
    return run_checks(checks)

File: examples/checks/complex_matrix_stability.go
package main

import (
	"math"
	"math/cmplx"
	"strings"
)

func checkComplexMatrix(ctx *Context) []Check {
	d := ctx.M()
	reported := map[string]struct {
		radius float64
		class  string
	}{}
	for _, m := range reAll(ctx.Answer, `(?m)^([A-Za-z0-9_]+) : spectral radius ([0-9.]+) -> ([^\n]+)$`) {
		reported[m[1]] = struct {
			radius float64
			class  string
		}{parseFloat(m[2]), strings.TrimSpace(m[3])}
	}
	classify := func(radius float64) string {
		if radius < 1.0 {
			return "damped"
		}
		if math.Abs(radius-1.0) <= 1e-12 {
			return "marginally stable"
		}
		return "unstable"
	}
	radii := map[string]float64{}
	for _, mat := range maps(d["matrices"]) {
		name := str(mat["name"])
		mx := 0.0
		for _, z := range maps(mat["diagonal"]) {
			r := math.Hypot(num(z["re"]), num(z["im"]))
			if r > mx {
				mx = r
			}
		}
		radii[name] = mx
	}
	rowsOK := len(reported) == len(radii)
	for name, radius := range radii {
		row, ok := reported[name]
		rowsOK = rowsOK && ok && math.Abs(row.radius-radius) <= 1e-12 && row.class == classify(radius)
	}
	zm := asMap(asMap(d["sampleProduct"])["z"])
	wm := asMap(asMap(d["sampleProduct"])["w"])
	z := complex(num(zm["re"]), num(zm["im"]))
	w := complex(num(wm["re"]), num(wm["im"]))
	scale := num(d["scale"])
	return []Check{{"diagonal entries are used as the eigenvalues", len(radii) == len(maps(d["matrices"]))}, {"A_unstable has independently recomputed spectral radius 2", math.Abs(radii["A_unstable"]-2.0) <= 1e-12 && classify(radii["A_unstable"]) == "unstable"}, {"A_stable has spectral radius exactly 1 and is marginal", math.Abs(radii["A_stable"]-1.0) <= 1e-12 && classify(radii["A_stable"]) == "marginally stable"}, {"A_damped has spectral radius 0 and is damped", radii["A_damped"] == 0.0 && classify(radii["A_damped"]) == "damped"}, {"reported matrix classes and radii match recomputation", rowsOK}, {"squared modulus of z*w equals product of squared moduli", math.Abs(math.Pow(cmplx.Abs(z*w), 2)-math.Pow(cmplx.Abs(z), 2)*math.Pow(cmplx.Abs(w), 2)) <= 1e-12}, {"scaling a matrix by 2 multiplies spectral-radius-squared by 4", math.Abs(math.Pow(scale*radii["A_unstable"], 2)-math.Pow(scale, 2)*math.Pow(radii["A_unstable"], 2)) <= 1e-12}}
}
